featurebyte/routes/lazy_app_container.py

```diff
@@ -0,0 +1,284 @@
+"""
+Lazy app container functions the same as the app_container, but only initializes dependencies when needed.
+"""
+from typing import Any, Dict, List
+
+from bson import ObjectId
+
+from featurebyte.persistent import Persistent
+from featurebyte.routes.app_container_config import AppContainerConfig, ClassDefinition, DepType
+from featurebyte.routes.task.controller import TaskController
+from featurebyte.routes.temp_data.controller import TempDataController
+from featurebyte.service.task_manager import TaskManager
+from featurebyte.storage import Storage
+from featurebyte.utils.credential import MongoBackedCredentialProvider
+
+
+def get_all_deps_for_key(
+    key: str, class_def_mapping: Dict[str, ClassDefinition], base_deps: Dict[str, Any]
+) -> List[str]:
+    """
+    Get dependencies for a given key.
+
+    Parameters
+    ----------
+    key: str
+        key to get dependencies for
+    class_def_mapping: Dict[str, ClassDefinition]
+        mapping of key to class definition
+    base_deps: Dict[str, Any]
+        existing dependencies
+
+    Returns
+    -------
+    Dict[str, ClassDefinition]
+        ordered dependencies, with the elements
+    """
+    all_deps = [key]
+    # Skip if the key is already in the base_deps
+    if key in base_deps:
+        return all_deps
+
+    # Get class definition
+    class_def = class_def_mapping[key]
+    dependencies = class_def.dependencies
+    # If this node has no children, return the current node.
+    if not dependencies:
+        return all_deps
+
+    # Recursively get all dependencies of the children
+    for dep in dependencies:
+        children_deps = get_all_deps_for_key(dep, class_def_mapping, base_deps)
+        children_deps.extend(all_deps)
+        all_deps = children_deps
+
+    return all_deps
```

Comment:
For this recursive function, we may need to update the base case logic by returning it earlier if it has been explored.

Consider following case:
```
A -> B  # A depends on B 
A -> C
B -> D
C -> D
D -> E
E -> F
F -> G
```
In this case, A depends on B & C, both B & C depends on D. If we use this function to find all dependencies, we should explore D node once rather than twice. For example:
```
A -> B -> D -> ...
   -> C -> check D has been explored and return (rather than keep exploring D's dependencies)
```

Reply:
good point - fixed